src/shared/notify-recv.c

```diff
@@ -63,44 +66,68 @@ int notify_recv(int fd, char **ret_text, struct ucred *ret_ucred, PidRef *ret_pi
 
                 case SCM_PIDFD:
                         assert(cmsg->cmsg_len == CMSG_LEN(sizeof(int)));
+                        assert(pidfd < 0);
+
                         pidfd = *CMSG_TYPED_DATA(cmsg, int);
                         break;
 
                 case SCM_CREDENTIALS:
                         assert(cmsg->cmsg_len == CMSG_LEN(sizeof(struct ucred)));
+                        assert(!ucred);
+
                         ucred = CMSG_TYPED_DATA(cmsg, struct ucred);
                         break;
                 }
         }
 
-        if ((ret_ucred || ret_pidref) && (!ucred || ucred->pid <= 0))
-                return log_debug_errno(SYNTHETIC_ERRNO(EAGAIN), "Got notification datagram lacking valid credential information, ignoring.");
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
 
-        if (n == 0)
-                return log_debug_errno(SYNTHETIC_ERRNO(EAGAIN), "Got empty notification message, ignoring.");
-        if (memchr(buf, 0, n - 1))
-                return log_debug_errno(SYNTHETIC_ERRNO(EAGAIN), "Got notification message with embedded NUL, ignoring.");
-
-        if (ret_pidref) {
-                assert(ucred);
-                assert(ucred->pid > 0);
+        if (ret_ucred || ret_pidref) {
+                if (!ucred || !pid_is_valid(ucred->pid))
+                        return log_warning_errno(SYNTHETIC_ERRNO(EAGAIN),
+                                                 "Got notification datagram lacking valid credential information, ignoring.");
 
                 if (pidfd >= 0)
-                        *ret_pidref = (PidRef) {
-                                .pid = ucred->pid,
-                                .fd = TAKE_FD(pidfd),
-                        };
+                        r = pidref_set_pidfd_consume(&pidref, TAKE_FD(pidfd));
                 else
-                        *ret_pidref = PIDREF_MAKE_FROM_PID(ucred->pid);
+                        r = pidref_set_pid(&pidref, ucred->pid);
```

Comment:
i really don't like this part of the change. previously we could receive messages from reaped processes and could handle them completely correctly. With this now we cannot, becase pidref_set_pidfd_consume() will fail if the process died by now. I think that's a really bad idea.

Hence, there's a reason why I did the explicit compound initialuation of the two fields: so this definitely works.

Reply:
I see that the old version in pid 1 in manager_dispatch_notify_fd() did it like this, but I think it's worse than the version in notify_recv() so far, for the aforementioned reasons.